### ISSUE TITLE
New "undo" patterns

### ISSUE THREAD

**Issue by paulwalk** [antleaf/notify-documentation#7]
Currently we have only one example of an `undo` pattern, _[Undo Offer](https://notify.coar-repositories.org/patterns/undo-offer/)_

In a recent discussion with PCI we identified a possible use for `undo` with regard to a previous `accept` activity. The workflow is something like:

1. Repository `offer`s a preprint for review to PCI
2. PCI sends notification of decision to (`accept`) this `offer`
3. PCI later decides that it will not, after all, review the preprint

I suggest that a good pattern to conclude this workflow is to send two notifications:
4. PCI sends `undo` relating to the `accept` in notification (3) above
5. PCI sends notification of decision to (`reject`) the `offer` made in (1)

This might seem a little redundant, but I think the intention is made clear and there is the possibility of an 'audit' record that is complete.

The semantics seem fine since the `undo` activity is meant to be applied to any other type of activity. ([docs](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-undo))

I have created two patterns - "undo accept" and "undo reject" in a new branch called `new_undo_patterns` in this repo. You can see this by checking out that branch and then running Hugo in the usual way. But here's a screen shot for convenience:

**Comment by paulwalk** [comment]
> An undo of a previous accept by itself is insufficient, IMO, as it leave the offer un-responded to.

I agree with that.

> A reject of the offer, IMO, voids previous accept notifications and is therefore sufficient.

I agree that this is sufficient, in the sense that the intended state can be *inferred* by comparing dates.

The reason I like the idea of adding an `undo` is that this makes a statement which does **not** require inference based on dates.

So, to be clear, I do suggest that the `undo` is necessary - but I think it could be good practice.

**Comment by hvdsomp** [comment]
I appreciate the perspectives of both @paulwalk and @martinklein0815:

- Sending two messages is redundant
- Sending only a single message requires intelligent handling of an Inbox in order to take notifications that were handled/finalized off the stack

I also appreciate the observation of @martinklein0815 that, if a single message is sent, it must be a `reject` of the `offer`. 

Combining these insights, it seems to me that a way to approach this scenario would be to state that:

1. a `reject` of the `offer` must be sent
2. an `undo` of the `accept` may be sent 

As per @paulwalk, there's no harm in (2). The only question that stating both (1) and (2) raises for me is whether a service node will remember which one it is they must send given there's two they could send. And whether they might send the wrong one. In which case, it might be safer to only state (1).

Having said all of this, and thinking aloud, I am wondering whether all this complication is really needed. Could the service node not just act as if all is cool and, in alignment with the `accept` of the `offer`, provide a service result that is not a review but just some kind of opaque statement? From the perspective of open science, it seems to me that lifecycle events for an artifact like the one at hand, i.e. a review was requested for this artifact from that service, is relevant for transparency. And, it seems that the proposed approach really tries to erase an episode from an artifact's lifecycle. Can't say I am a fan ...

**Comment by paulwalk** [comment]
Just on this point:

> And, it seems that the proposed approach really tries to erase an episode from an artifact's lifecycle. Can't say I am a fan 

I am also not very enthusiastic about this aspect... but it is PCI's policy not to advertise a refused endorsement.

Having said that, other services may face the workflow indicated in this issue but be open to announcing a refused endorsement. In such a case, it would be better to have something like the "opaque statement" that you suggest above.

I wonder if we could define a very generic notification that simply means "this activity was concluded", which could be used anywhere that any `activity` should simply be halted? This might even actually work better for PCI's case too.

**Comment by phochste** [comment]
There are three activities we didn't yet consider:

- [TentativeAccept](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-tentativeaccept)
- [TentativeReject](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-tentativereject)

and maybe

- [Ignore](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-ignore)

**Comment by martinklein0815** [comment]
Thanks, @phochste! TentativeAccept is interesting and could be used as a response by PCI to the initial offer. The "tentative" notion might be valuable as PCI requires users/authors to fill out a form once they receive an offer but before they start the review process. If the user does not do that, a TentativeAccept could expire or repositories could time out their review request (offer). If a user completes this step, PCI could follow up with a "real" Accept. This flow helps distinguish the case where a user does not fulfill PCI requirements (form) from the case where PCI's reviews are not positive and the manuscript is not endorsed. 

I am not sure TentativeAccept helps for the scenario where PCI's reviews are not positive.

**Comment by paulwalk** [comment]
I agree with this too. `TentativeAccept`, followed by an `Accept` or `Reject` seems to achieve the right semantics for PCI's intent in the earlier part of the workflow.

(I just don't get where `Ignore` would be used to be honest - if you're sending a notification about something then you're not ignoring it...)

In the later stage, when PCI wants to halt the workflow with no reason given, I don't think we have a good solution. I had initially though this might be a sort of generic `Announce` but announce` is always related to an `Object`, so that doesn't really work.

So, the general question is: how to halt an exchange of notifications without leaving the other party "hanging" (i.e. unsure if some activity is ongoing)?

**Comment by phochste** [comment]
Typically in ActivityPub one doesn't send messages to stop a conversation. 

What could be an option is to [Leave](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-leave) with the original Offer as object.

Or use [Flag](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-flag) the Offer with a COAR specific sub activity ?

**Comment by paulwalk** [comment]
When I said:

>  I had initially though this might be a sort of generic Announce but Announce is always related to an Object, so that doesn't really work.

I now think I was rejecting this idea too hastily. Actually, an `Announce` could work if the `Object` was the original `Offer`. We would just need to define a new type - as we have for example for `[EndorsementAction](https://notify.coar-repositories.org/vocabulary/EndorsementAction/)` - perhaps `HaltingAction` or similar.

So, in the concrete example for PCI, if the process has got to the stage where the reviews have actually happened, and the outcome is not to endorse, then PCI sends a notification with:

`Object`: the original `Offer`
`Type`: `Announce` & `HaltingAction`

Alternatively, we could use `Flag` instead of `Announce`, although I'm uneasy about the semantics of `Flag` which I think might imply that the original `Offer` was somehow inappropriate (which is not true)

**Comment by paulwalk** [comment]
I'm really not sure what we should suggest. From the point of view of the protocol, the best solution is for PCI to send a notification indicating that the review has concluded and that endorsement was not forthcoming. But this is what their policy does not allow, so the repository is left "hanging", not knowing if the process has finished or not.

I think `Flag` is problematic, because it has connotations of reporting inappropriate activity, which is not the case here.

`Leave` feels a bit of a stretch... but maybe... I don't have any better idea!

**Comment by paulwalk** [comment]
I think there are are actually two situations we have been trying to address, and that we have a good solution to the first of these.

### Situation 1
The `Offer` is accepted, but then the process does not go ahead for some reason (perhaps because the author does not supply enough information). PCI decides to stop the process early. No review has taken place, so the situation is *not* that an endorsement has been refused based on poor reviews.

**Solution to situation 1:** Use `TenatativeAccept` for the original acknowledgement. If the process needs to be stopped before the reviews are begun, then issue a `Reject` acknowledgment. If  the process can continue into the review stage, then issue an `Accept` acknowledgement. I think with this arrangement there is no need for an `Undo`.


### Situation 2
The process has progressed past the `Accept` stage and reviews have been conducted. The review indicate that no endorsement should be given.

**Solution to situation 2:** I think this is where the "opaque document" comes in, but I confess I'm not entirely sure what this looks like...?

> Another aspect to consider is whether the exception scenario under discussion should be published as an actual Notify pattern and hence be promoted? Or could it just be a dialect extension of Notify used by a bunch of buddies?

I think we should publish any *pattern* which is used, but we certainly don't need to document all *scenarios* (collections of patterns implementing some workflow).

**Comment by hvdsomp** [comment]
Situation 2 is what this issue is about. Situation 1 was introduced as part of the discussion of this issue.

I don’t know what the opaque document looks like either. But, as indicated above, I assume that somehow, pre-Notify, they’re also informing the requester of an endorsement that it’s not going to happen. Whatever they tell the requester is what they could say in the opaque document. If they don’t tell the requester anything, pre-Notify, and just let him/her hanging there, why would things have to be done differently under Notify?

**Comment by hvdsomp** [comment]
The opaque document is a Service Result. So, from the perspective of Notify, it's just a URL in a notification sent in response to the offer that was accepted. Turns out that the document at that URL is not a not of endorsement (as would be the case if all is OK) but rather an note stating "I am afraid we can't help you" (because all is not OK).

**Comment by paulwalk** [comment]
Ok, thanks. So possibly an ‘Announce’ pattern without any other more specific ‘Type’, and an ‘Object’ that has the URI to the document.

As you said in a previous comment, the visibility of the notification exchange is significant, since the existence of a failed bid for endorsement can be extrapolated from all this. But I feel like that aspect is out of the scope of our concern.

**Comment by paulwalk** [comment]
I have done the following:

1. Added two new Notify patterns for `TentativeAccept` and `TentativeReject`
2. Added a generic `Announcement` pattern called "Announce Service Result"

These are now on the [live documentation](https://notify.coar-repositories.org/patterns/).

I will explain all this to PCI next.

**Comment by phochste** [comment]
Hi there are still typos in the `TentativeAccept` and `TentativeReject` . The example and the type says `Accept` and `Reject` tehre

**Comment by paulwalk** [comment]
Doh! Thanks @phochste - fixed now.
